### DISCUSSION TITLE
Create a model and mainfest for serverless functions

### DISCUSSION THREAD

**Discussion by navveenb** [Green-Software-Foundation/hack#32] [Develop an Idea]
## Story
As a user, I want to calculate the environmental impact of serverless functions

## Rationale
Serverless computing is a cloud-based model where the cloud provider manages server resources, enabling developers to write and deploy code without worrying about the underlying infrastructure. This idea is to generate a manifest file and model that could be used to calculate the energy and carbon consumed by serverless functions. 

## Scope
This is a manifest file. See [here](https://github.com/Green-Software-Foundation/if/tree/dev/examples/impls/test) for examples and [here](https://if.greensoftware.foundation/tutorials/how-to-write-impls) for a manifest writing tutorial.

## Implementation guidelines
* The manifest MUST be a single .yaml file
* The manifest file MUST have all the required fields described [here](https://if.greensoftware.foundation/tutorials/how-to-write-impls#structure-of-an-impl)
* The manifest MUST define the necessary inputs required for the calculation
* The manifest MUST include all necessary models in a pipeline
* The manifest MUST define the expected outputs and their units
* The manifest SHOULD include comments showing which models are necessary but currently unavailable
* The manifest SHOULD include some documentation in a separate README explaining the execution flow, and any known problems or missing components.

## Difficulty
3/5

## Impact
3/5

## Examples and resources
[IMPL writing guide](https://if.greensoftware.foundation/tutorials/how-to-write-impls)

**Comment by josh-swerdlow** [comment]
Navveen (or anyone reading)

I was really interested by this proposal as a possible importer plugin, but ultimately my co-contributor and I decided on something else. I'm still interested and am wondering if you can expand on how this can be implemented? The main stopper for me was a lack of knowledge on how or if this was possible. I'm new to serverless and am not familiar with aws, azure, gcp offerings.

My quick perusal of aws and azure found that they have extensive monitoring solutions for their cloud offerings that can probably give one CPU utilization per process or VM. I imagine these could extend if your enterprise has dedicated resources to house your serverless functions as they scale; however, I don't know if monitoring would even be enabled for some random person's serverless use case. My understanding is that a random persons serverless function is deployed on whatever resources are available at the moment. Perhaps this is based on some user constraints, but I would be doubtful if the observability solutions are enabled.

Do you think you one would use some already provided observability tooling/API to fetch environment level aggregation of metrics? If so do you know what metrics are exposed?

If not that, then do you think one could write a library that developers could include in their serverless functions to perform calls to the relevant hardware to collect these metrics? Afterward either export those to some aggregator or even directly include the IF to calculate some metric in the serverless function (though this seems to add bulk to the function that devs may want to avoid)?

Thanks!

**Comment by navveenb** [comment]
@josh-swerdlow  
The serverless methodology would be based on some approximation methodology.
For reference, please look at one of the Methodologies from Cloud Carbon Footprint (https://www.cloudcarbonfootprint.org/docs/methodology). (Look at AWS Lambda and AWS Aurora Serverless for reference.)
Similarly, the methodology can be assumed for other serverless components, based on certain assumptions.

**Comment by josh-swerdlow** [comment]
Thanks! That's really helpful and the project is more clear now. I'll circle back to this after the hackathon, but keep it in mind as I work on my hackathon plugin.
